File: implementations/php/pcf-sig/src/Verify.php
<?php

declare(strict_types=1);

namespace Kduma\PCFSIG;

use Kduma\PCF\Container;
use Kduma\PCF\PartitionEntry;

/** Verdict on one SignedEntry inside a Manifest (spec Section 11, V7). */
enum EntryVerdict: string
{
    case Valid = 'Valid';
    case MissingPartition = 'MissingPartition';
    case ProtectedFieldMismatch = 'ProtectedFieldMismatch';
    case DataHashRecomputationMismatch = 'DataHashRecomputationMismatch';
    case WeakHash = 'WeakHash';
}

/** Verdict on a whole PCFSIG_SIG partition (spec Section 11, V8). */
enum ManifestVerdict: string
{
    case Valid = 'Valid';
    case Invalid = 'Invalid';
    case Unverifiable = 'Unverifiable';
}

/** Why a manifest could not be verified. */
enum UnverifiableReason: string
{
    case NoMatchingKey = 'NoMatchingKey';
    case UnsupportedSigAlgo = 'UnsupportedSigAlgo';
    case UnsupportedKeyFormat = 'UnsupportedKeyFormat';
    case MalformedKey = 'MalformedKey';
    case SignatureLengthMismatch = 'SignatureLengthMismatch';
}

/** Per-entry report. */
final class EntryReport
{
    public function __construct(
        public readonly string $uid,
        public EntryVerdict $verdict,
    ) {
    }
}

/** Report for one PCFSIG_SIG partition. */
final class SignatureReport
{
    /** @param EntryReport[] $entries */
    public function __construct(
        public readonly string $sigPartitionUid,
        public readonly string $signerKeyFingerprint,
        public readonly int $signedAtUnixSeconds,
        public ManifestVerdict $verdict,
        public ?UnverifiableReason $unverifiableReason,
        public ?int $unverifiableId,
        public array $entries,
    ) {
    }
}

/** Whether to independently re-hash each covered partition during verification. */
enum DataRecheck: string
{
    case Skip = 'Skip';
    case Recompute = 'Recompute';
}

/** High-level verification API (spec Section 11). */
final class Verify
{
    /**
     * Verify every PCFSIG_SIG partition in $container and return one report
     * each.
     *
     * @return SignatureReport[]
     */
    public static function all(
        Container $container,
        DataRecheck $recheck = DataRecheck::Skip,
    ): array {
        $entries = $container->entries();

        /** @var array<array{record: KeyRecord, uid: string}> $keys */
        $keys = [];
        foreach ($entries as $e) {
            if ($e->partitionType === Consts::TYPE_PCFSIG_KEY) {
                try {
                    $rec = KeyRecord::fromBytes($container->readPartitionData($e));
                    $keys[] = ['record' => $rec, 'uid' => $e->uid];
                } catch (PcfSigException) {
                    // skip
                }
            }
        }

        /** @var SignatureReport[] $reports */
        $reports = [];
        foreach ($entries as $e) {
            if ($e->partitionType !== Consts::TYPE_PCFSIG_SIG) {
                continue;
            }
            $data = $container->readPartitionData($e);
            $reports[] = self::verifyOne($entries, $keys, $e, $data);
        }

        if ($recheck === DataRecheck::Recompute) {
            foreach ($reports as $r) {
                foreach ($r->entries as $er) {
                    if ($er->verdict !== EntryVerdict::Valid) {
                        continue;
                    }
                    $p = null;
                    foreach ($entries as $x) {
                        if ($x->uid === $er->uid) {
                            $p = $x;
                            break;
                        }
                    }
                    if ($p !== null) {
                        $bytes = $container->readPartitionData($p);
                        $computed = $p->dataHashAlgo->compute($bytes);
                        if (!hash_equals($computed, $p->dataHash)) {
                            $er->verdict = EntryVerdict::DataHashRecomputationMismatch;
                        }
                    }
                }
            }
        }

        return $reports;
    }

    /** @return SignatureReport[] */
    public static function allWithRecheck(Container $container): array
    {
        return self::all($container, DataRecheck::Recompute);
    }

    /**
     * @param PartitionEntry[]                                $entries
     * @param array<array{record: KeyRecord, uid: string}>    $keys
     */
    private static function verifyOne(
        array $entries,
        array $keys,
        PartitionEntry $sigEntry,
        string $data,
    ): SignatureReport {
        try {
            $parsed = SignaturePartition::fromBytes($data);
        } catch (PcfSigException) {
            return new SignatureReport(
                $sigEntry->uid,
                str_repeat("\x00", Consts::FINGERPRINT_SIZE),
                0,
                ManifestVerdict::Unverifiable,
                UnverifiableReason::MalformedKey,
                null,
                [],
            );
        }

        $report = new SignatureReport(
            $sigEntry->uid,
            $parsed->manifest->signerKeyFingerprint,
            $parsed->manifest->signedAtUnixSeconds,
            ManifestVerdict::Valid,
            null,
            null,
            [],
        );

        // Self-reference check (spec Section 7.2).
        foreach ($parsed->manifest->signedEntries as $e) {
            if ($e->uid === $sigEntry->uid) {
                $report->verdict = ManifestVerdict::Invalid;

                return $report;
            }
        }

        if (!$parsed->manifest->sigAlgo->isImplemented()) {
            $report->verdict = ManifestVerdict::Unverifiable;
            $report->unverifiableReason = UnverifiableReason::UnsupportedSigAlgo;
            $report->unverifiableId = $parsed->manifest->sigAlgo->id();

            return $report;
        }

        $key = null;
        foreach ($keys as $k) {
            if (hash_equals(
                $k['record']->fingerprint,
                $parsed->manifest->signerKeyFingerprint,
            )) {
                $key = $k;
                break;
            }
        }
        if ($key === null) {
            $report->verdict = ManifestVerdict::Unverifiable;
            $report->unverifiableReason = UnverifiableReason::NoMatchingKey;

            return $report;
        }

        $keyRecord = $key['record'];
        if (!$keyRecord->keyFormat->isImplemented()) {
            $report->verdict = ManifestVerdict::Unverifiable;
            $report->unverifiableReason = UnverifiableReason::UnsupportedKeyFormat;
            $report->unverifiableId = $keyRecord->keyFormat->id();

            return $report;
        }

        if (
            $parsed->manifest->sigAlgo === SigAlgo::Ed25519
            && $keyRecord->keyFormat === KeyFormat::Ed25519Raw
        ) {
            if (\strlen($parsed->signature) !== Consts::ED25519_SIGNATURE_LEN) {
                $report->verdict = ManifestVerdict::Unverifiable;
                $report->unverifiableReason = UnverifiableReason::SignatureLengthMismatch;

                return $report;
            }
            if (\strlen($keyRecord->keyData) !== Consts::ED25519_PUBLIC_KEY_LEN) {
                $report->verdict = ManifestVerdict::Unverifiable;
                $report->unverifiableReason = UnverifiableReason::MalformedKey;

                return $report;
            }
            try {
                $ok = sodium_crypto_sign_verify_detached(
                    $parsed->signature,
                    $parsed->manifestBytes,
                    $keyRecord->keyData,
                );
            } catch (\SodiumException) {
                $ok = false;
            }
            if (!$ok) {
                $report->verdict = ManifestVerdict::Invalid;

                return $report;
            }
        } else {
            $report->verdict = ManifestVerdict::Unverifiable;
            $report->unverifiableReason = UnverifiableReason::UnsupportedSigAlgo;
            $report->unverifiableId = $parsed->manifest->sigAlgo->id();

            return $report;
        }

        foreach ($parsed->manifest->signedEntries as $se) {
            $p = null;
            foreach ($entries as $x) {
                if ($x->uid === $se->uid) {
                    $p = $x;
                    break;
                }
            }
            if ($p === null) {
                $report->entries[] = new EntryReport($se->uid, EntryVerdict::MissingPartition);

                continue;
            }
            if (!Manifest::isCryptoHash($se->dataHashAlgo)) {
                $report->entries[] = new EntryReport($se->uid, EntryVerdict::WeakHash);

                continue;
            }
            if (
                $p->partitionType !== $se->partitionType
                || $p->label !== $se->label
                || $p->usedBytes !== $se->usedBytes
                || $p->dataHashAlgo !== $se->dataHashAlgo
                || !hash_equals($p->dataHash, $se->dataHash)
            ) {
                $report->entries[] = new EntryReport(
                    $se->uid,
                    EntryVerdict::ProtectedFieldMismatch,
                );

                continue;
            }
            $report->entries[] = new EntryReport($se->uid, EntryVerdict::Valid);
        }

        return $report;
    }
}

File: implementations/dotnet/pcf-sig/src/Pcf.Sig/Verify.cs
using System.Collections.Generic;
using Org.BouncyCastle.Math.EC.Rfc8032;
using Pcf;

namespace Pcf.Sig;

/// <summary>Verdict on one SignedEntry inside a Manifest (spec Section 11, V7).</summary>
public enum EntryVerdict
{
    /// <summary>Covered partition exists, all protected fields match, hash is cryptographic.</summary>
    Valid,
    /// <summary>No partition in the container has the SignedEntry's uid.</summary>
    MissingPartition,
    /// <summary>A protected field of the live partition does not match the manifest.</summary>
    ProtectedFieldMismatch,
    /// <summary>Recomputed digest of live partition data does not match the SignedEntry's data_hash.</summary>
    DataHashRecomputationMismatch,
    /// <summary>The covered partition's data_hash_algo_id is not cryptographic.</summary>
    WeakHash,
}

/// <summary>Verdict on a whole PCFSIG_SIG partition (spec Section 11, V8).</summary>
public enum ManifestVerdict
{
    /// <summary>Manifest parsed; signature cryptographically verified against the referenced key.</summary>
    Valid,
    /// <summary>Manifest parsed; signature did NOT verify against the referenced key.</summary>
    Invalid,
    /// <summary>Manifest parsed but cannot be verified (no matching key, or unsupported alg/format).</summary>
    Unverifiable,
}

/// <summary>Why a manifest could not be verified.</summary>
public enum UnverifiableReason
{
    /// <summary>No PCFSIG_KEY partition with the manifest's signer_key_fingerprint.</summary>
    NoMatchingKey,
    /// <summary>The signature algorithm id is not implemented by this build.</summary>
    UnsupportedSigAlgo,
    /// <summary>The key format id is not implemented by this build.</summary>
    UnsupportedKeyFormat,
    /// <summary>The matching key partition is malformed.</summary>
    MalformedKey,
    /// <summary>The signature byte length does not match the algorithm's natural size.</summary>
    SignatureLengthMismatch,
}

/// <summary>Per-entry report.</summary>
public sealed class EntryReport
{
    /// <summary>The SignedEntry's uid.</summary>
    public byte[] Uid { get; }

    /// <summary>Verdict for this entry.</summary>
    public EntryVerdict Verdict { get; set; }

    /// <summary>Construct a per-entry report.</summary>
    public EntryReport(byte[] uid, EntryVerdict verdict)
    {
        Uid = uid;
        Verdict = verdict;
    }
}

/// <summary>Report for one PCFSIG_SIG partition.</summary>
public sealed class SignatureReport
{
    /// <summary>PCF uid of the PCFSIG_SIG partition itself.</summary>
    public byte[] SigPartitionUid { get; set; }

    /// <summary><c>signer_key_fingerprint</c> copied from the manifest.</summary>
    public byte[] SignerKeyFingerprint { get; set; }

    /// <summary><c>signed_at_unix_seconds</c> copied from the manifest.</summary>
    public long SignedAtUnixSeconds { get; set; }

    /// <summary>Verdict on the manifest as a whole.</summary>
    public ManifestVerdict Verdict { get; set; }

    /// <summary>Detailed reason when <see cref="Verdict"/> is <see cref="ManifestVerdict.Unverifiable"/>.</summary>
    public UnverifiableReason? UnverifiableReason { get; set; }

    /// <summary>Optional id detail (e.g., unsupported algorithm id).</summary>
    public int? UnverifiableId { get; set; }

    /// <summary>Per-entry verdicts.</summary>
    public List<EntryReport> Entries { get; set; } = new();
}

/// <summary>Whether to independently re-hash each covered partition during verification.</summary>
public enum DataRecheck
{
    /// <summary>Trust the PCF data_hash field as captured by the SignedEntry.</summary>
    Skip,
    /// <summary>Recompute hash(partition bytes) and compare to the SignedEntry's data_hash.</summary>
    Recompute,
}

/// <summary>High-level verification API (spec Section 11).</summary>
public static class Verify
{
    /// <summary>Verify every PCFSIG_SIG partition and return one report each.</summary>
    public static List<SignatureReport> All(
        Container container,
        DataRecheck recheck = DataRecheck.Skip)
    {
        var entries = container.Entries();

        var keys = new List<(KeyRecord Record, byte[] Uid)>();
        foreach (var e in entries)
        {
            if (e.PartitionType == Constants.TypePcfsigKey)
            {
                try
                {
                    var rec = KeyRecord.FromBytes(container.ReadPartitionData(e));
                    keys.Add((rec, e.Uid));
                }
                catch (PcfSigException)
                {
                    // skip malformed
                }
            }
        }

        var reports = new List<SignatureReport>();
        foreach (var e in entries)
        {
            if (e.PartitionType != Constants.TypePcfsigSig) continue;
            var data = container.ReadPartitionData(e);
            reports.Add(VerifyOne(entries, keys, e, data));
        }

        if (recheck == DataRecheck.Recompute)
        {
            foreach (var r in reports)
            {
                foreach (var er in r.Entries)
                {
                    if (er.Verdict != EntryVerdict.Valid) continue;
                    PartitionEntry p = null;
                    foreach (var x in entries)
                    {
                        if (BytesEqual(x.Uid, er.Uid)) { p = x; break; }
                    }
                    if (p != null)
                    {
                        var bytes = container.ReadPartitionData(p);
                        var computed = p.DataHashAlgo.Compute(bytes);
                        if (!BytesEqual(computed, p.DataHash))
                        {
                            er.Verdict = EntryVerdict.DataHashRecomputationMismatch;
                        }
                    }
                }
            }
        }

        return reports;
    }

    /// <summary>Same as <see cref="All"/> with <see cref="DataRecheck.Recompute"/>.</summary>
    public static List<SignatureReport> AllWithRecheck(Container container) =>
        All(container, DataRecheck.Recompute);

    private static SignatureReport VerifyOne(
        List<PartitionEntry> entries,
        List<(KeyRecord Record, byte[] Uid)> keys,
        PartitionEntry sigEntry,
        byte[] data)
    {
        SignaturePartition parsed;
        try
        {
            parsed = SignaturePartition.FromBytes(data);
        }
        catch (PcfSigException)
        {
            return new SignatureReport
            {
                SigPartitionUid = sigEntry.Uid,
                SignerKeyFingerprint = new byte[Constants.FingerprintSize],
                SignedAtUnixSeconds = 0,
                Verdict = ManifestVerdict.Unverifiable,
                UnverifiableReason = Pcf.Sig.UnverifiableReason.MalformedKey,
            };
        }

        var report = new SignatureReport
        {
            SigPartitionUid = sigEntry.Uid,
            SignerKeyFingerprint = parsed.Manifest.SignerKeyFingerprint,
            SignedAtUnixSeconds = parsed.Manifest.SignedAtUnixSeconds,
            Verdict = ManifestVerdict.Valid,
        };

        foreach (var e in parsed.Manifest.SignedEntries)
        {
            if (BytesEqual(e.Uid, sigEntry.Uid))
            {
                report.Verdict = ManifestVerdict.Invalid;
                return report;
            }
        }

        if (!parsed.Manifest.SigAlgo.IsImplemented())
        {
            report.Verdict = ManifestVerdict.Unverifiable;
            report.UnverifiableReason = Pcf.Sig.UnverifiableReason.UnsupportedSigAlgo;
            report.UnverifiableId = (byte)parsed.Manifest.SigAlgo;
            return report;
        }

        (KeyRecord Record, byte[] Uid)? key = null;
        foreach (var k in keys)
        {
            if (BytesEqual(k.Record.Fingerprint, parsed.Manifest.SignerKeyFingerprint))
            {
                key = k;
                break;
            }
        }
        if (key == null)
        {
            report.Verdict = ManifestVerdict.Unverifiable;
            report.UnverifiableReason = Pcf.Sig.UnverifiableReason.NoMatchingKey;
            return report;
        }

        if (!key.Value.Record.KeyFormat.IsImplemented())
        {
            report.Verdict = ManifestVerdict.Unverifiable;
            report.UnverifiableReason = Pcf.Sig.UnverifiableReason.UnsupportedKeyFormat;
            report.UnverifiableId = (byte)key.Value.Record.KeyFormat;
            return report;
        }

        if (parsed.Manifest.SigAlgo == SigAlgo.Ed25519
            && key.Value.Record.KeyFormat == KeyFormat.Ed25519Raw)
        {
            if (parsed.Signature.Length != Constants.Ed25519SignatureLen)
            {
                report.Verdict = ManifestVerdict.Unverifiable;
                report.UnverifiableReason = Pcf.Sig.UnverifiableReason.SignatureLengthMismatch;
                return report;
            }
            if (key.Value.Record.KeyData.Length != Constants.Ed25519PublicKeyLen)
            {
                report.Verdict = ManifestVerdict.Unverifiable;
                report.UnverifiableReason = Pcf.Sig.UnverifiableReason.MalformedKey;
                return report;
            }
            bool ok;
            try
            {
                ok = Ed25519.Verify(
                    parsed.Signature, 0,
                    key.Value.Record.KeyData, 0,
                    parsed.ManifestBytes, 0, parsed.ManifestBytes.Length);
            }
            catch
            {
                ok = false;
            }
            if (!ok)
            {
                report.Verdict = ManifestVerdict.Invalid;
                return report;
            }
        }
        else
        {
            report.Verdict = ManifestVerdict.Unverifiable;
            report.UnverifiableReason = Pcf.Sig.UnverifiableReason.UnsupportedSigAlgo;
            report.UnverifiableId = (byte)parsed.Manifest.SigAlgo;
            return report;
        }

        foreach (var se in parsed.Manifest.SignedEntries)
        {
            PartitionEntry p = null;
            foreach (var x in entries)
            {
                if (BytesEqual(x.Uid, se.Uid)) { p = x; break; }
            }
            EntryVerdict verdict;
            if (p == null)
            {
                verdict = EntryVerdict.MissingPartition;
            }
            else if (!Manifest.IsCryptoHash(se.DataHashAlgo))
            {
                verdict = EntryVerdict.WeakHash;
            }
            else if (p.PartitionType != se.PartitionType
                     || !BytesEqual(p.Label, se.Label)
                     || p.UsedBytes != se.UsedBytes
                     || p.DataHashAlgo != se.DataHashAlgo
                     || !BytesEqual(p.DataHash, se.DataHash))
            {
                verdict = EntryVerdict.ProtectedFieldMismatch;
            }
            else
            {
                verdict = EntryVerdict.Valid;
            }
            report.Entries.Add(new EntryReport(se.Uid, verdict));
        }

        return report;
    }

    private static bool BytesEqual(byte[] a, byte[] b)
    {
        if (a.Length != b.Length) return false;
        for (int i = 0; i < a.Length; i++) if (a[i] != b[i]) return false;
        return true;
    }
}
